src/api/api.js

```diff
@@ -36,78 +35,85 @@ api_schema.register_api(require('./func_api'));
-        this.node = undefined;
-        this.host = undefined;
-        this.bucket = undefined;
-        this.events = undefined;
-        this.object = undefined;
-        this.agent = undefined;
-        this.block_store = undefined;
-        this.stats = undefined;
-        this.scrubber = undefined;
-        this.debug = undefined;
-        this.redirector = undefined;
-        this.tiering_policy = undefined;
-        this.pool = undefined;
-        this.cluster_server = undefined;
-        this.cluster_internal = undefined;
-        this.server_inter_process = undefined;
-        this.hosted_agents = undefined;
-        this.frontend_notifications = undefined;
-        this.func = undefined;
-        this.func_node = undefined;
-
-        _.each(rpc.schema, api => {
-            if (!api || !api.id || api.id[0] === '_') return;
-            const name = api.id.replace(/_api$/, '');
-            if (name === 'rpc' || name === 'options') throw new Error('ILLEGAL API ID');
-            this[name] = {};
-            _.each(api.methods, (method_api, method_name) => {
-                this[name][method_name] = (params, options) => {
-                    options = _.create(this.options, options);
-                    return rpc._request(api, method_api, params, options);
-                };
-            });
-        });
-    }
+const shared_client_proto = {
+    RPC_BUFFERS,
 
-    /**
-     * extend the rpc client prototype with convinient methods
-     */
-    create_auth_token(params) {
-        return this.auth.create_auth(params)
-            .then(res => {
-                this.options.auth_token = res.token;
-                return res;
-            });
-    }
+    create_auth_token: async function(params) {
+        const res = this.auth.create_auth(params);
+        this.options.auth_token = res.token;
+        return res;
+    },
 
-    create_access_key_auth(params) {
-        return this.auth.create_access_key_auth(params)
-            .then(res => {
-                this.options.auth_token = res.token;
-                return res;
-            });
-    }
+    create_access_key_auth: async function(params) {
+        const res = this.auth.create_access_key_auth(params);
+        this.options.auth_token = res.token;
+        return res;
+    },
 
-    async create_k8s_auth(params) {
+    create_k8s_auth: async function(params) {
         const res = await this.auth.create_k8s_auth(params);
         this.options.auth_token = res.token;
         return res;
+    },
+
+    _invoke_api(api, method_api, params, options) {
+        options = Object.assign(
+            Object.create(this.options),
+            options
+        );
+        return this.rpc._request(api, method_api, params, options);
     }
+};
 
+function new_client_factory(rpc) {
+    // Copy the shared part instead of using it as a proto
+    // to skip one hoop when invoking these methods.
+    const client_proto = {
+        ...shared_client_proto,
+        rpc
```

Comment:
Perhaps we can leave this rpc property undefined and the rpc itself will set the rpc property to this after the creation of the client? This way we can compile this factory method just from the schema.

Reply:
We can but then this will make every client proto twice as big (it will hold the options object and the RPC reference.

The fact is that the RPC is shared for all these instances which make sense to put it on the prototype

Reply:
The size different is negligible.
I don't think this goes against the change as we discussed it this change is mostly to allow scavenging, which a reference to rpc will not change.

Reply:
Another thing on readability - I think the separation of shared_client_proto and the client_proto feels a bit weird because in any case this is called just once to create a client_proto, and shared_client_proto is very small, so why not just create it once inside new_client_factory?

Reply:
About the rpc reference, I still think the rpc reference should be added to the client_proto, so that client objects are not even effected in size, right?

Reply:
I agree that if we remove the RPC form the proto then we can generate the factory only once and not once per RPC instance. In this case, we can merge the shard and actual proto.
The separation was done because in the current case we have a new client_porto for each RPC instance.

Your last comment does not make sense to me. Changing the proto of an object after creation is not recommended performance-wise. It will cause decompilation of all code that uses instances of objects that use this proto.
Also, this will bring us back to a client_proto per RPC instance which you advocated against in your comments

Reply:
I don't suggest changing the proto after creating objects from it.
I meant before - when RPC ctor is running.

Reply:
You can't you have one proto for many RPCs

Reply:
I decoupled the API from the RPC (adding the RPC reference to the client).